engine/table/src/main/java/io/deephaven/engine/table/impl/sources/UnionSourceManager.java

```diff
@@ -5,365 +5,647 @@
+     * iterates the data structure during construction and merged notification delivery, with one exception:
+     * {@link MergedUnionListener#canExecute(long)}, which is mutually-synchronized with all modification operations.
+     */
+    private final IntrusiveDoublyLinkedQueue<LinkedListenerRecorder> listenerRecorders;
+    private final MergedListener mergedListener;
+    private final ConstituentChangesListenerRecorder constituentChangesListener;
+    private final UpdateCommitter<UnionSourceManager> updateCommitter;
+
+    public UnionSourceManager(@NotNull final PartitionedTable partitionedTable) {
+        constituentChangesPermitted = partitionedTable.constituentChangesPermitted();
+        columnNames = partitionedTable.constituentDefinition().getColumnNamesArray();
+
+        final Table coalescedPartitions = partitionedTable.table().coalesce().select(
+                new TableTransformationColumn(partitionedTable.constituentColumnName(), Table::coalesce));
+        constituentRows = coalescedPartitions.getRowSet();
+        constituentTables = coalescedPartitions.getColumnSource(partitionedTable.constituentColumnName());
+
+        final boolean refreshing = coalescedPartitions.isRefreshing();
+        final int initialNumSlots = constituentRows.intSize();
+
+        // noinspection resource
+        resultRows = RowSetFactory.empty().toTracking();
+        unionRedirection = new UnionRedirection(initialNumSlots, refreshing);
+        // noinspection unchecked
+        resultColumnSources = partitionedTable.constituentDefinition().getColumnStream()
+                .map(cd -> new UnionColumnSource<>(cd.getDataType(), cd.getComponentType(), this, unionRedirection,
+                        new TableSourceLookup(cd.getName())))
                 .toArray(UnionColumnSource[]::new);
-        names = tableDefinition.getColumnList().stream().map(ColumnDefinition::getName).toArray(String[]::new);
-        this.parentDependency = parentDependency;
-
-        result = new QueryTable(RowSetFactory.empty().toTracking(), getColumnSources());
-        rowSet = result.getRowSet().writableCast();
-        modifiedColumnSet = result.newModifiedColumnSet(names);
+        resultTable = new QueryTable(resultRows, getColumnSources());
+        modifiedColumnSet = resultTable.getModifiedColumnSetForUpdates();
 
-        mergedListener = new MergedUnionListener(listeners, "TableTools.merge()", result);
-    }
+        if (refreshing) {
+            listenerRecorders = new IntrusiveDoublyLinkedQueue<>(
+                    IntrusiveDoublyLinkedNode.Adapter.<LinkedListenerRecorder>getInstance());
+            mergedListener = new MergedUnionListener(listenerRecorders, resultTable);
+            resultTable.addParentReference(mergedListener);
+
+            constituentChangesListener = new ConstituentChangesListenerRecorder(coalescedPartitions);
```

Comment:
Should we be listening for constituent changes if we have a static list of constituents, as in the case of TableTools.merge(a, b)?

Reply:
I thought it best to listen and verify:
```
    private TableUpdate getAndCheckConstituentChanges() {
        final TableUpdate constituentChanges = constituentChangesListener.getUpdate();
        if (!constituentChangesPermitted && constituentChanges != null && !constituentChanges.empty()) {
            throw new IllegalStateException(
                    "Constituent changes not permitted, but received update " + constituentChanges);
        }
        return constituentChanges == null ? EMPTY_TABLE_UPDATE : constituentChanges;
    }
```
I don't actually know the provenance of the table at this point; that is, was it built from an array, or did some user create it?